### ISSUE TITLE
Implementation of `numpy.linalg.eigvals` equivalent

### ISSUE THREAD

**Issue by N0rbert** [cupy/cupy#6359]
### Description

NumPy has a [`numpy.linalg.eigvals`](https://numpy.org/doc/stable/reference/generated/numpy.linalg.eigvals.html) function, which is often used in real-life programs.
Please provide an equivalent of it for CuPy and name it `cupy.linalg.eigvals`.

**Comment by emcastillo** [comment]
CuPy relies on `cuSolver` for implementing these linear algebra routines. However, there are no routines in cuSolver for non-symmetric matrices, so this requires implementing Cuda kernels or custom routines to solve this.

**Comment by leofang** [comment]
I think we might be able to steal MAGMA's kernels for this purpose 😄

**Comment by emcastillo** [comment]
yeah, I njust checked and actually torch uses MAGMA for this!

**Comment by IvanYashchuk** [comment]
I'm sorry to disappoint but MAGMA's eigenvalue solver for non-symmetric matrices is really bad. There's no specialized GPU-only path, it uses a mix of LAPACK calls with a tiny bit of CUDA kernels. Eigenvalues are computed using LAPACK.
See lots of `lapack77` calls in https://bitbucket.org/icl/magma/src/656d51d82375e791f3f6b20f47271cdbc41d5a8a/src/dgeev.cpp

Up to about 3k x 3k matrices, it's faster to compute on CPU.

**Comment by emcastillo** [comment]
Thanks @IvanYashchuk!!

I was looking into MAGMA code and saw the very same as you, I am just thinking of naively throwing this into a QR solver despite being slow just to avoid CPU sync.

**Comment by emcastillo** [comment]
@N0rbert @IvanYashchuk hi

We have discussed this internally, and some people suggest that it is better to use `svd` if possible. And if there is a case in your applications where `svd` can't be used and you really need `eigvals` no matter what.

Thanks

**Comment by FedorIVS** [comment]
It seems to me that `cuSolver` now has the necessary implementations for non-symmetric matrices:

https://docs.nvidia.com/cuda/cusolver/index.html#cusolverdn-dense-lapack

If I'm correct, could this request be revisited?

Thanks!

**Comment by kmaehashi** [comment]
cuSOLVER does not cover this yet. See https://docs.nvidia.com/cuda/cusolver/index.html#cusolverdnxsyevd

**Comment by byron158** [comment]
It looks like the latest cusolver has an implementation for non-symmetric matrices. Is it possible to revisit this request? (Looks like it was added in cuSOLVER: Release 12.6 Update 2)
https://docs.nvidia.com/cuda/cusolver/index.html#cusolverdnxgeev

**Comment by kmaehashi** [comment]
> It looks like the latest cusolver has an implementation for non-symmetric matrices. Is it possible to revisit this request? (Looks like it was added in cuSOLVER: Release 12.6 Update 2) https://docs.nvidia.com/cuda/cusolver/index.html#cusolverdnxgeev

Thanks for the info, @byron158! The docs says:

> The following routine:
> ...
> computes for an n-by-n real non-symmetric or complex non-Hermitian matrix `A` the eigenvalues and, optionally, the left and/or right eigenvectors.

**Comment by kmaehashi** [comment]
xref #8729, another request

**Comment by leofang** [comment]
Yes! The new routine was meant for CuPy and other projects to implement `eig` and `eigvals`.

**Comment by timzsu** [comment]
Hi, 

This topic sounds quite interesting to me. Is there anyone already working on this? If not I would like to work on this (and #3255 as well) :)

BR

**Comment by atykhyy** [comment]
> It looks like the latest cusolver has an implementation for non-symmetric matrices. Is it possible to revisit this request? (Looks like it was added in cuSOLVER: Release 12.6 Update 2) https://docs.nvidia.com/cuda/cusolver/index.html#cusolverdnxgeev

Remark 3 in the docs for `cusolverDnXgeev` says that it is "a hybrid CPU-GPU algorithm" with "best performance attained with pinned host memory", which leads me to suspect that cuSOLVER uses the same sort of code as MAGMA, with key steps of `geev` (reduction to upper Hessenberg form and the QR algorithm) not being GPU-accelerated. (Also there is no `cusolverDnXggev` to solve generalized non-symmetric/non-Hermitian eigenvalue problems.) Therefore this new routine is unlikely to achieve the kind of orders-of-magnitude performance improvement we get for symmetric/Hermitian eigenproblems. There are a few papers on GPU acceleration of the upper Hessenberg reduction, all many years old, none of which seem to have resulted in widely adopted implementations. This is regrettable, but perhaps not too surprising, as university departments which need to solve large non-symmetric/non-Hermitian eigenproblems tend to have access to compute clusters and need not side-track their scarce resources into developing GPU implementations.

**Comment by codecircuit** [comment]
Hi atykhyy,

chiming in here, as I was one of the contributors for the implementation of cusolverDnXgeev.

> which leads me to suspect that cuSOLVER uses the same sort of code as MAGMA

cusolverDnXgeev is a different implementation and the key steps are GPU accelerated. With eigenvector computation, I would expect a few x speedup compared to Magma.

**Comment by atykhyy** [comment]
@codecircuit I installed the most recent CUDA on my Windows machine (not the newest, a Thinkpad P52 with a discrete NVidia GPU Quadro P3200 with Max-Q design, CUDA compute version 6.1), installed cupy from the PR referenced in this thread, and tried the solver on a non-symmetric 10000x10000 matrix with a simple `w,v=cupy.linalg.eig(A)`. I can see that the solver is indeed GPU-accelerated - task manager showed the GPU's CUDA engine chugging at 100% using all 6Gb dedicated GPU RAM - and the solver found the same eigenvalues and eigenvectors of interest to me, but unexpectedly it took 23'40" to complete where scipy's CPU-only solver using default settings (`w,v=scipy.linalg.eig(A)`) took 9'50" on the exact same matrix. Is this too large a matrix size for this amount of GPU RAM, or is my hardware too old, or what's going on?

**Comment by codecircuit** [comment]
@atykhyy could it be that you do the computation in FP64? Your graphics card has very poor FP64 performance. Switching to FP32 should speed up the computation but has an effect on the convergence of the algorithm and could result into more iterations. You can see the number of iterations and used precision by enabling cuSOLVER logs by setting environment variable `CUSOLVERDN_LOG_MASK=18`.
